test/unit/Cardano/Wallet/BlockSyncerSpec.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE DataKinds #-}
+module Cardano.Wallet.BlockSyncerSpec
+    ( spec
+    ) where
+
+
+import Prelude
+
+
+import Control.Concurrent
+    ( forkIO, killThread, threadDelay, throwTo )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Data.ByteString
+    ( ByteString, pack )
+import Data.IORef
+    ( newIORef, readIORef )
+import qualified Data.Set as Set
+import System.Exit
+    ( ExitCode (..) )
+import Test.Hspec
+    ( Spec, describe, it )
+import Test.Hspec.Expectations
+    ( shouldBe )
+import Test.QuickCheck
+    ( generate, vector )
+import Test.QuickCheck.Gen
+    ( Gen )
+
+
+import Cardano.Wallet.BlockSyncer
+    ( BlocksConsumed (..), blockSyncerBracket, runBlockLayerMockup )
+import Cardano.Wallet.Primitive
+    ( Block (..), BlockHeader (..), Hash (..) )
+
+
+spec :: Spec
+spec = do
+    describe "Block syncer downloads all blocks" $ do
+        it "the bracket releases itself as expected when exception is thrown" $ do
+            consecutiveBlocks <- liftIO $ mkConsecutiveTestBlocks 1
+            blockLayerMockup <- runBlockLayerMockup 2 consecutiveBlocks 2
+
+            initialHeaderData <- newIORef $ BlocksConsumed []
+            threadID <- forkIO $ blockSyncerBracket initialHeaderData blockLayerMockup 2
+            threadDelay (1*1000*1000)
+            throwTo threadID ExitSuccess
+
+        it "the bracket releases itself as expected when its thread is killed" $ do
+            consecutiveBlocks <- liftIO $ mkConsecutiveTestBlocks 1
+            blockLayerMockup <- runBlockLayerMockup 2 consecutiveBlocks 2
+
+            initialHeaderData <- newIORef $ BlocksConsumed []
+            threadID <- forkIO $ blockSyncerBracket initialHeaderData blockLayerMockup 2
+            threadDelay (1*1000*1000)
+            killThread threadID
+
+        it "when the block polling's ticking function is 5x more frequent" $ do
+            testBlocksPolling 10 1 5
+        it "when the block polling's ticking function is 2x more frequent" $ do
+            testBlocksPolling 10 1 2
+        it "when the block polling's ticking function and mockup ticking is are the same" $ do
+            testBlocksPolling 20 1 1
+        it "when the mockup ticking is 5x more frequent" $ do
+            testBlocksPolling 10 5 1
+        it "when the mockup ticking is 2x more frequent" $ do
+            testBlocksPolling 10 2 1
+
+    where
+        testBlocksPolling
+            :: Int
+            -> Int
+            -> Int
```

Comment:
In such case, think about using a record type to make the intent clear about what the parameters are.

Reply:
:-)